File: js/src/game.ts
import { draw_ui, fetch_json, find_html_element_by_id, hide_ui } from "./browser";
import { Audio, IGame, KeyState, Renderer, load_image, Image, SpriteSheet, add_click_handler, Rect } from "./engine";
import { IObstacle } from "./obstacles";
import { RedHatBoy } from "./red_hat_boy";
import { platform_and_stone, stone_and_platform } from "./segments";

const TIMELINE_MINIMUM: number = 1000;
const OBSTACLE_BUFFER: number = 20;

function rightmost(obstacle_list: Array<IObstacle>): number {
  return obstacle_list.map((obstacle) => obstacle.right()).reduce((a, b) => (b > a ? b : a), 0);
}

interface IWalk {
  obstacle_sheet: SpriteSheet;
  boy: RedHatBoy;
  backgrounds: [Image, Image];
  obstacles: Array<IObstacle>;
  stone: HTMLImageElement;
  timeline: number;
}

class Walk implements IWalk {
  constructor(
    public boy: RedHatBoy,
    public backgrounds: [Image, Image],
    public obstacles: Array<IObstacle>,
    public obstacle_sheet: SpriteSheet,
    public stone: HTMLImageElement,
    public timeline: number
  ) {}

  velocity(): number {
    return -this.boy.walking_speed();
  }

  generate_next_segment(): void {
    const next_segment = Math.floor(Math.random() * 2);
    let next_obstacles: Array<IObstacle>;
    switch (next_segment) {
      case 0:
        next_obstacles = stone_and_platform(this.stone, this.obstacle_sheet, this.timeline + OBSTACLE_BUFFER);
        break;
      case 1:
        next_obstacles = platform_and_stone(this.stone, this.obstacle_sheet, this.timeline + OBSTACLE_BUFFER);
        break;
      default:
        next_obstacles = [];
    }
    this.timeline = rightmost(next_obstacles);
    this.obstacles.push(...next_obstacles);
  }

  draw(renderer: Renderer): void {
    this.backgrounds.forEach((background) => {
      background.draw(renderer);
    });
    this.boy.draw(renderer);
    this.obstacles.forEach((obstacle) => {
      obstacle.draw(renderer);
    });
  }

  knocked_out(): boolean {
    return this.boy.knocked_out();
  }

  static reset(walk: Walk): Walk {
    const starting_obstacles = stone_and_platform(walk.stone, walk.obstacle_sheet, 0);
    const timeline = rightmost(starting_obstacles);
    return new Walk(
      RedHatBoy.reset(walk.boy),
      walk.backgrounds,
      starting_obstacles,
      walk.obstacle_sheet,
      walk.stone,
      timeline
    );
  }
}

enum WalkTheDogStateType {
  Ready,
  Walking,
  GameOver,
}

interface IWalkTheDogState {
  type: WalkTheDogStateType;
  walk: Walk;
  draw(renderer: Renderer): void;
}

class WalkingWalkTheDogState implements IWalkTheDogState {
  type = WalkTheDogStateType.Walking;
  constructor(public walk: Walk) {}

  update(keystate: KeyState): WalkingWalkTheDogState | GameOverWalkTheDogState {
    if (keystate.is_pressed("ArrowDown")) {
      this.walk.boy.slide();
    }
    if (keystate.is_pressed("Space")) {
      this.walk.boy.jump();
    }
    this.walk.boy.update();
    let walking_speed = this.walk.velocity();
    let [first_background, second_background] = this.walk.backgrounds;
    first_background.move_horizontally(walking_speed);
    second_background.move_horizontally(walking_speed);
    if (first_background.right() < 0) {
      first_background.set_x(second_background.right());
    }
    if (second_background.right() < 0) {
      second_background.set_x(first_background.right());
    }
    this.walk.obstacles = this.walk.obstacles.filter((obstacle) => obstacle.right() > 0);
    this.walk.obstacles.forEach((obstacle) => {
      obstacle.move_horizontally(walking_speed);
      obstacle.check_intersection(this.walk.boy);
    });
    if (this.walk.timeline < TIMELINE_MINIMUM) {
      this.walk.generate_next_segment();
    } else {
      this.walk.timeline += walking_speed;
    }
    if (this.walk.knocked_out()) {
      return this.end_game();
    } else {
      return this;
    }
  }

  end_game(): GameOverWalkTheDogState {
    draw_ui("<button id='new_game'>New Game</button");
    const el = find_html_element_by_id("new_game");
    const nextState = new GameOverWalkTheDogState(this.walk);
    add_click_handler(el, () => (nextState.new_game_pressed = true));
    return nextState;
  }

  draw(renderer: Renderer): void {
    this.walk.draw(renderer);
  }
}

class ReadyWalkTheDogState implements IWalkTheDogState {
  type = WalkTheDogStateType.Ready;
  constructor(public walk: Walk) {}

  update(keystate: KeyState): ReadyWalkTheDogState | WalkingWalkTheDogState {
    this.walk.boy.update();
    if (keystate.is_pressed("ArrowRight")) {
      return this.start_running();
    } else {
      return this;
    }
  }

  start_running(): WalkingWalkTheDogState {
    this.run_right();
    return new WalkingWalkTheDogState(this.walk);
  }

  run_right(): void {
    this.walk.boy.run_right();
  }

  draw(renderer: Renderer): void {
    this.walk.draw(renderer);
  }
}

class GameOverWalkTheDogState implements IWalkTheDogState {
  type = WalkTheDogStateType.GameOver;
  public new_game_pressed: boolean = false;

  constructor(public walk: Walk) {}

  update(): GameOverWalkTheDogState | ReadyWalkTheDogState {
    if (this.new_game_pressed) {
      return this.new_game();
    } else {
      return this;
    }
  }

  new_game(): ReadyWalkTheDogState {
    hide_ui();
    return new ReadyWalkTheDogState(Walk.reset(this.walk));
  }

  draw(renderer: Renderer): void {
    this.walk.draw(renderer);
  }
}

class WalkTheDogStateMachine {
  private state: IWalkTheDogState;

  constructor(walk: Walk) {
    this.state = new ReadyWalkTheDogState(walk);
  }

  update(keystate: KeyState): WalkTheDogStateMachine {
    if (this.state.type === WalkTheDogStateType.Ready) {
      this.state = (<ReadyWalkTheDogState>this.state).update(keystate);
    } else if (this.state.type === WalkTheDogStateType.Walking) {
      this.state = (<WalkingWalkTheDogState>this.state).update(keystate);
    } else if (this.state.type === WalkTheDogStateType.GameOver) {
      this.state = (<GameOverWalkTheDogState>this.state).update();
    }
    return this;
  }

  draw(renderer: Renderer): void {
    this.state.draw(renderer);
  }
}

interface IWalkTheDog {
  machine: WalkTheDogStateMachine | undefined;
}

export class WalkTheDog implements IWalkTheDog, IGame {
  machine: WalkTheDogStateMachine | undefined;

  async initialize(): Promise<IGame> {
    if (this.machine === undefined) {
      let json = await fetch_json("rhb.json");
      let background = await load_image("BG.png");
      let background_width = background.width;
      let stone = await load_image("Stone.png");
      let tiles_json = await fetch_json("tiles.json");
      let tiles = new SpriteSheet(tiles_json, await load_image("tiles.png"));
      let starting_obstacles = stone_and_platform(stone, tiles, 0);
      let timeline = rightmost(starting_obstacles);
      let audio = new Audio();
      let sound = await audio.load_sound("SFX_Jump_23.mp3");
      let background_music = await audio.load_sound("background_song.mp3");
      audio.play_looping_sound(background_music);
      let rhb = new RedHatBoy(json, await load_image("rhb.png"), audio, sound);
      this.machine = new WalkTheDogStateMachine(
        new Walk(
          rhb,
          [
            new Image(background, { x: 0, y: 0 }),
            new Image(background, {
              x: background_width,
              y: 0,
            }),
          ],
          starting_obstacles,
          tiles,
          stone,
          timeline
        )
      );
      setTimeout(() => {});
    } else {
      throw new Error("State machine already exists");
    }
    return this;
  }

  update(keystate: KeyState): void {
    if (this.machine !== undefined) {
      this.machine = this.machine.update(keystate);
    }
  }

  draw(renderer: Renderer): void {
    renderer.clear(Rect.new_from_x_y(0, 0, 600, 570));
    if (this.machine !== undefined) {
      this.machine.draw(renderer);
    }
  }
}

File: rust/src/game.rs
use std::rc::Rc;

use crate::engine::{
    load_image, Audio, Game, Image, KeyState, Point, Rect, Renderer, Sheet, SpriteSheet,
};
use crate::obstacles::Obstacle;
use crate::red_hat_boy::RedHatBoy;
use crate::segments::{platform_and_stone, stone_and_platform};
use crate::{browser, engine};
use anyhow::{anyhow, Result};
use async_trait::async_trait;
use futures::channel::mpsc::UnboundedReceiver;
use rand::{thread_rng, Rng};
use web_sys::HtmlImageElement;

const TIMELINE_MINIMUM: i16 = 1000;
const OBSTACLE_BUFFER: i16 = 20;

fn rightmost(obstacle_list: &[Box<dyn Obstacle>]) -> i16 {
    obstacle_list
        .iter()
        .map(|obstacle| obstacle.right())
        .max_by(|x, y| x.cmp(&y))
        .unwrap_or(0)
}

pub struct Walk {
    obstacle_sheet: Rc<SpriteSheet>,
    boy: RedHatBoy,
    backgrounds: [Image; 2],
    obstacles: Vec<Box<dyn Obstacle>>,
    stone: HtmlImageElement,
    timeline: i16,
}

impl Walk {
    fn velocity(&self) -> i16 {
        -self.boy.walking_speed()
    }

    fn generate_next_segment(&mut self) {
        let mut rng = thread_rng();
        let next_segment = rng.gen_range(0..2);
        let mut next_obstacles = match next_segment {
            0 => stone_and_platform(
                self.stone.clone(),
                self.obstacle_sheet.clone(),
                self.timeline + OBSTACLE_BUFFER,
            ),
            1 => platform_and_stone(
                self.stone.clone(),
                self.obstacle_sheet.clone(),
                self.timeline + OBSTACLE_BUFFER,
            ),
            _ => vec![],
        };
        self.timeline = rightmost(&next_obstacles);
        self.obstacles.append(&mut next_obstacles);
    }

    fn draw(&self, renderer: &Renderer) {
        self.backgrounds.iter().for_each(|background| {
            background.draw(renderer);
        });
        self.boy.draw(renderer);
        self.obstacles.iter().for_each(|obstacle| {
            obstacle.draw(renderer);
        });
    }

    fn knocked_out(&self) -> bool {
        self.boy.knocked_out()
    }

    fn reset(walk: Self) -> Self {
        let starting_obstacles =
            stone_and_platform(walk.stone.clone(), walk.obstacle_sheet.clone(), 0);
        let timeline = rightmost(&starting_obstacles);

        Walk {
            boy: RedHatBoy::reset(walk.boy),
            backgrounds: walk.backgrounds,
            obstacles: starting_obstacles,
            obstacle_sheet: walk.obstacle_sheet,
            stone: walk.stone,
            timeline,
        }
    }
}

enum ReadyEndState {
    Complete(WalkTheDogState<Walking>),
    Continue(WalkTheDogState<Ready>),
}

impl From<ReadyEndState> for WalkTheDogStateMachine {
    fn from(state: ReadyEndState) -> Self {
        match state {
            ReadyEndState::Complete(walking) => walking.into(),
            ReadyEndState::Continue(ready) => ready.into(),
        }
    }
}

struct WalkTheDogState<T> {
    _state: T,
    walk: Walk,
}

impl<T> WalkTheDogState<T> {
    fn draw(&self, renderer: &Renderer) {
        self.walk.draw(renderer);
    }
}

enum WalkingEndState {
    Complete(WalkTheDogState<GameOver>),
    Continue(WalkTheDogState<Walking>),
}

impl From<WalkingEndState> for WalkTheDogStateMachine {
    fn from(state: WalkingEndState) -> Self {
        match state {
            WalkingEndState::Complete(game_over) => game_over.into(),
            WalkingEndState::Continue(walking) => walking.into(),
        }
    }
}

enum GameOverEndState {
    Complete(WalkTheDogState<Ready>),
    Continue(WalkTheDogState<GameOver>),
}

impl From<GameOverEndState> for WalkTheDogStateMachine {
    fn from(state: GameOverEndState) -> Self {
        match state {
            GameOverEndState::Complete(ready) => ready.into(),
            GameOverEndState::Continue(game_over) => game_over.into(),
        }
    }
}

struct Ready;
struct Walking;
struct GameOver {
    new_game_event: UnboundedReceiver<()>,
}

impl GameOver {
    fn new_game_pressed(&mut self) -> bool {
        matches!(self.new_game_event.try_next(), Ok(Some(())))
    }
}

impl WalkTheDogState<Ready> {
    fn update(mut self, keystate: &KeyState) -> ReadyEndState {
        self.walk.boy.update();
        if keystate.is_pressed("ArrowRight") {
            ReadyEndState::Complete(self.start_running())
        } else {
            ReadyEndState::Continue(self)
        }
    }

    fn start_running(mut self) -> WalkTheDogState<Walking> {
        self.run_right();
        WalkTheDogState {
            _state: Walking,
            walk: self.walk,
        }
    }

    fn run_right(&mut self) {
        self.walk.boy.run_right();
    }

    fn new(walk: Walk) -> WalkTheDogState<Ready> {
        WalkTheDogState {
            _state: Ready,
            walk,
        }
    }
}

impl WalkTheDogState<Walking> {
    fn update(mut self, keystate: &KeyState) -> WalkingEndState {
        if keystate.is_pressed("ArrowDown") {
            self.walk.boy.slide();
        }
        if keystate.is_pressed("Space") {
            self.walk.boy.jump();
        }
        self.walk.boy.update();
        let walking_speed = self.walk.velocity();
        let [first_background, second_background] = &mut self.walk.backgrounds;
        first_background.move_horizontally(walking_speed);
        second_background.move_horizontally(walking_speed);
        if first_background.right() < 0 {
            first_background.set_x(second_background.right());
        }
        if second_background.right() < 0 {
            second_background.set_x(first_background.right());
        }
        self.walk.obstacles.retain(|obstacle| obstacle.right() > 0);
        self.walk.obstacles.iter_mut().for_each(|obstacle| {
            obstacle.move_horizontally(walking_speed);
            obstacle.check_intersection(&mut self.walk.boy);
        });
        if self.walk.timeline < TIMELINE_MINIMUM {
            self.walk.generate_next_segment()
        } else {
            self.walk.timeline += walking_speed;
        }
        if self.walk.knocked_out() {
            WalkingEndState::Complete(self.end_game())
        } else {
            WalkingEndState::Continue(self)
        }
    }

    fn end_game(self) -> WalkTheDogState<GameOver> {
        let receiver = browser::draw_ui("<button id='new_game'>New Game</button")
            .and_then(|_unit| browser::find_html_element_by_id("new_game"))
            .map(|element| engine::add_click_handler(element))
            .unwrap();
        WalkTheDogState {
            _state: GameOver {
                new_game_event: receiver,
            },
            walk: self.walk,
        }
    }
}

impl WalkTheDogState<GameOver> {
    fn update(mut self) -> GameOverEndState {
        if self._state.new_game_pressed() {
            GameOverEndState::Complete(self.new_game())
        } else {
            GameOverEndState::Continue(self)
        }
    }

    fn new_game(self) -> WalkTheDogState<Ready> {
        browser::hide_ui();
        WalkTheDogState {
            _state: Ready,
            walk: Walk::reset(self.walk),
        }
    }
}

impl From<WalkTheDogState<Ready>> for WalkTheDogStateMachine {
    fn from(state: WalkTheDogState<Ready>) -> Self {
        WalkTheDogStateMachine::Ready(state)
    }
}

impl From<WalkTheDogState<Walking>> for WalkTheDogStateMachine {
    fn from(state: WalkTheDogState<Walking>) -> Self {
        WalkTheDogStateMachine::Walking(state)
    }
}

impl From<WalkTheDogState<GameOver>> for WalkTheDogStateMachine {
    fn from(state: WalkTheDogState<GameOver>) -> Self {
        WalkTheDogStateMachine::GameOver(state)
    }
}

enum WalkTheDogStateMachine {
    Ready(WalkTheDogState<Ready>),
    Walking(WalkTheDogState<Walking>),
    GameOver(WalkTheDogState<GameOver>),
}

impl WalkTheDogStateMachine {
    fn update(self, keystate: &KeyState) -> Self {
        match self {
            WalkTheDogStateMachine::Ready(state) => state.update(keystate).into(),
            WalkTheDogStateMachine::Walking(state) => state.update(keystate).into(),
            WalkTheDogStateMachine::GameOver(state) => state.update().into(),
        }
    }

    fn draw(&self, renderer: &Renderer) {
        match self {
            WalkTheDogStateMachine::Ready(state) => state.draw(renderer),
            WalkTheDogStateMachine::Walking(state) => state.draw(renderer),
            WalkTheDogStateMachine::GameOver(state) => state.draw(renderer),
        }
    }

    fn new(walk: Walk) -> Self {
        WalkTheDogStateMachine::Ready(WalkTheDogState::new(walk))
    }
}

pub struct WalkTheDog {
    machine: Option<WalkTheDogStateMachine>,
}

impl WalkTheDog {
    pub fn new() -> Self {
        WalkTheDog { machine: None }
    }
}

#[async_trait(?Send)]
impl Game for WalkTheDog {
    async fn initialize(&self) -> Result<Box<dyn Game>> {
        match self.machine {
            None => {
                let json = browser::fetch_json("rhb.json").await?;
                let audio = Audio::new()?;
                let sound = audio.load_sound("SFX_Jump_23.mp3").await?;
                let background_music = audio.load_sound("background_song.mp3").await?;
                audio.play_looping_sound(&background_music);
                let rhb = RedHatBoy::new(
                    json.into_serde::<Sheet>()?,
                    load_image("rhb.png").await?,
                    audio,
                    sound,
                );
                let background = load_image("BG.png").await?;
                let background_width = background.width() as i16;
                let stone = load_image("Stone.png").await?;
                let tiles_json = browser::fetch_json("tiles.json").await?;
                let tiles = Rc::new(SpriteSheet::new(
                    tiles_json.into_serde::<Sheet>()?,
                    load_image("tiles.png").await?,
                ));
                let starting_obstacles = stone_and_platform(stone.clone(), tiles.clone(), 0);
                let timeline = rightmost(&starting_obstacles);
                let machine = WalkTheDogStateMachine::new(Walk {
                    obstacle_sheet: tiles,
                    boy: rhb,
                    backgrounds: [
                        Image::new(background.clone(), Point { x: 0, y: 0 }),
                        Image::new(
                            background.clone(),
                            Point {
                                x: background_width,
                                y: 0,
                            },
                        ),
                    ],
                    obstacles: starting_obstacles,
                    stone,
                    timeline,
                });
                Ok(Box::new(WalkTheDog {
                    machine: Some(machine),
                }))
            }
            Some(_) => Err(anyhow!("Error: Game is already initialized!")),
        }
    }

    fn update(&mut self, keystate: &KeyState) {
        if let Some(machine) = self.machine.take() {
            self.machine.replace(machine.update(keystate));
        }
        assert!(self.machine.is_some());
    }

    fn draw(&self, renderer: &Renderer) {
        renderer.clear(&Rect::new_from_x_y(0, 0, 600, 570));
        if let Some(machine) = &self.machine {
            machine.draw(renderer);
        }
    }
}
